libmamba/src/core/package_handling.cpp

```diff
@@ -443,46 +474,87 @@ namespace mamba
                 throw std::runtime_error(archive_error_string(ext));
             }
         }
-        archive_read_close(a);
-        archive_read_free(a);
+
         archive_write_close(ext);
         archive_write_free(ext);
 
         fs::current_path(prev_path);
     }
 
+
+    static la_ssize_t file_read(struct archive* a, void* client_data, const void** buff)
+    {
+        struct conda_extract_context* mine = static_cast<conda_extract_context*>(client_data);
+        *buff = mine->buffer.data();
+
+        auto read = archive_read_data(mine->source, mine->buffer.data(), mine->buffer.size());
+        if (read < 0)
+        {
+            throw std::runtime_error(
+                fmt::format("Error reading from archive: {}", archive_error_string(mine->source)));
+        }
+        return read;
+    }
+
+    int archive_read_open_archive_entry(struct archive* a, conda_extract_context* ctx)
+    {
+        archive_clear_error(a);
+        archive_read_set_read_callback(a, file_read);
+        archive_read_set_callback_data(a, ctx);
+        return archive_read_open1(a);
+    }
+
+
     void extract_conda(const fs::u8path& file,
                        const fs::u8path& dest_dir,
                        const std::vector<std::string>& parts)
     {
-        TemporaryDirectory tdir;
-        extract_archive(file, tdir);
+        // open outer zip archive
+        struct archive* a = archive_read_new();
+        struct archive_entry* entry;
```

Comment:
`struct` unnecessary and these needs RAII/scoping (`mamba::on_scope_exit`) as there are potential exceptions below.